snafu/image_pull_wrapper/trigger_image_pull.py

```diff
@@ -0,0 +1,119 @@
+        self.user = args.user
+        self.pod_name = args.pod_name
+        self.pod_count = args.pod_count
+        self.timeout = args.timeout
+        self.retries = args.retries
+        self.image = args.image
+
+    def _json_payload(self, data):
+        payload = {
+            "uuid": self.uuid,
+            "cluster_name": self.cluster_name,
+            "pod_name": self.pod_name,
+            "timeout": self.timeout,
+            "pod_count": self.pod_count,
+            "user": self.user,
+            "image_retries": self.retries,
+            "image": self.image,
+        }
+        payload.update(data)
+        return payload
+
+    def _run_image_pull_test(self):
+        logger.info("Running: Podman Pull on supplied image")
+
+        try:
+            if not os.path.exists("/tmp/image_pull"):
+                os.mkdir("/tmp/image_pull")
+        except OSError as err:
+            logger.error("Errored while trying to create /tmp/image_pull directory")
+            logger.error(err)
+            exit(1)
+
+        results = {}
+        cmd = ["skopeo", "--insecure-policy", "copy", self.image, "dir:/tmp/image_pull"]
+
+        failures = 0
+        while failures <= self.retries:
+            # Time the Push
+            start_time = datetime.datetime.utcnow()
+            p = Popen(cmd, stdout=PIPE, stderr=PIPE)
+            output, errors = p.communicate()
+            end_time = datetime.datetime.utcnow()
+
+            # Handle Errors
+            try:
+                assert p.returncode == 0
+                success = True
+            except Exception:
+                success = False
+                failures = failures + 1
+                logger.info("Failed to pull image: %s" % self.image)
+                logger.info("STDOUT: %s" % output)
+                logger.info("STDERR: %s" % errors)
+                logger.info("Retrying. %s/%s failures." % (failures, self.retries))
+
+            # Statistics / Data
+            elapsed_time = end_time - start_time
+            data = {
+                "image": self.image,
+                "elapsed_time": elapsed_time.total_seconds(),
+                "start_time": start_time,
+                "end_time": end_time,
+                "failures": failures,
+                "successful": success,
+            }
+            results.update(data)
+
+            if success:
+                break
+        try:
+            shutil.rmtree("/tmp/image_pull")
+        except OSError as err:
+            logger.error("Errored while trying to delete /tmp/image_pull directory")
+            logger.error(err)
+        return results
+
+    def emit_actions(self):
+        logger.info(
+            "Running image pull test with %d image retries and image: %s" % (self.retries, str(self.image))
+        )
+        logger.info("Test UUID is %s on cluster %s" % (self.uuid, self.cluster_name))
+
+        timestamp = time.strftime("%Y-%m-%dT%H:%M:%S")
```

Comment:
It might be useful to have a common, custom JSONEncoder for all of the tests wrapped by snafu.